### ISSUE TITLE
haproxy 2.2 breaks quoting/escaping single quotes

### ISSUE THREAD

**Issue by tmccombs** [haproxy/haproxy#1398]
### Detailed Description of the Problem

In haproxy 2.0 if I want to have a fetch that give sthe string constant `'self'`, I could use `str("'self'")` or `str(\'self\')` or even `str("\'self\'")`. However, in 2.2 both of these result in just `self`. The only way I've been able to get the desired `'self'` on 2.2 is with `str("\\'self\\'")` or `str(\"\'self\'\")`.  However, neither of these work in `2.0` which means, I can't have a single configuration that works for both versions.

### Expected Behavior

`str("'self'")`, `str(\'self\')`, and `str("\'self\'")`, should produce the string `'self'` instead of `self`, as was the case in older versions of haproxy.

### Steps to Reproduce the Behavior

1. Use the config given below
2. Make a request to haproxy
3. Look at the contents of the "test" response header.

### Do you have any idea what may have caused this?

I suspect this was introduced by the change in how haproxy parses arguments to fetches: https://github.com/haproxy/haproxy/commit/cd0d2ed6ee932e0d86b8fbaf5fcb97d9530bf260.

### What is your configuration?

```haproxy
defaults
    timeout client 10s
    timeout connect 2s
    timeout server 10s
    mode http

frontend test
  bind :8998

  http-request set-var(txn.test)  str("\'self\'")

  use_backend BACKEND_test

backend BACKEND_test
  http-after-response add-header text %[var(txn.test)]
  http-request return status 200

# vim: ft=haproxy
```

### Output of `haproxy -vv`

```plain
HA-Proxy version 2.2.17-dd94a25 2021/09/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.2.17.html
Running on: Linux 5.11.0-36-generic #40~20.04.1-Ubuntu SMP Sat Sep 18 02:14:19 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-stringop-overflow -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2_JIT=1 USE_LIBCRYPT=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT -PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=12).
Built with OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
Running on OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.34 2019-11-21
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.3.0
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
              h2 : mode=HTTP       side=FE|BE     mux=H2
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[COMP] compression
	[TRACE] trace
	[CACHE] cache
	[FCGI] fcgi-app

```

### Additional Information

Maybe this is the intended behavior with the changes to the parser? If so maybe the documentation for quoting and escaping should mention how to deal with single/strong quotes in keyword arguments, and that it is breaking change.

However, given the comment ["Good news guys, I managed to do this, while staying backwards compatible"](https://github.com/haproxy/haproxy/issues/200#issuecomment-586405106), I suspect this behavior wasn't intended to be a breaking change. Although, I'm not sure how feasible it is to fix.

**Comment by wtarreau** [comment]
The old behavior did not support nested quotes, nor did it support including delimiter characters in strings. For example it was not possible to include a comma in a string. Typically the following would refuse to parse:
```
   http-request redirect location "%[str('self(1,2)')]"
```
The reason was that the parser was a left->right parser so it would stop on delimiters, thus `'self(1` was the first argument to `str`, `2` was a second argument, and `')` was extra text.

With the rework that you pointed above, we addressed this by keeping the compatibility with all cases that previously worked normally (except those that already relied on nested quoting). The parser is now recursive like in normal languages. The first level of quotes and backslashes is handled by the line tokenizer, which is why when you enter `str("\\'self\\'")` what it does is pass `\'self\'` as the argument to `str`. This results in escaping the single quotes with the backslash so that it is not used as a string delimiter for the argument but as a character part of the string. Version 2.0 was unable to deal with this. And just because it was not possible to pass a delimiter to version 2.0 I can't find a way to have a same config working on both, I'm sorry.

If you need to simplify a migration, one option would be to set a global quote variable that you reuse elsewhere:
```
global
     # For haproxy >=2.2
     setenv Q "\\'"
     # for haproxy <=2.0
     #setenv Q "'"
...
  http-request set-var(txn.self) str("${Q}self${Q}")
  http-request redirect location %[var(txn.self)]
```

That's not perfect but it could allow you to have a single line to comment/uncomment (or you could even pass it externally based on a separate file or by the launcher script).

**Comment by tmccombs** [comment]
In the section on "quoting and escaping" in configuration.txt it states:

> The keyword argument parser is exactly the same as the top-level one regarding
> quotes, except that is will not make special cases of backslashes.

This was pretty confusing to me.  Apparently, what is meant is that the some escape sequences, like `\#`, `\xXX`, and `\$` aren't treated specially at the second layer but others like `\\` and `\'` are. 

I'll see if I can write a patch to try and make the configration.txt more clear.

**Comment by wtarreau** [comment]
I agree that "not make special case of backslashes" is very confusing, I'm sorry about this :-(  I wanted to write a BNF describing the language format but being constantly busy on other stuff I could not. I think it could be useful to some users, or maybe we'll figure it's too difficult to write one. Anyway I'd welcome a patch if you have any ideas. Thanks!

**Comment by capflam** [comment]
fixed by cd34ad713
